syntax = "proto3";

package employees.v1;

option go_package = "github.com/opensourcecorp/workshops/protobuf-grpc/pb/employees/v1";
// TODO:
// option java_package = "org.opensourcecorp.workshops.protobuf_grpc.employees.v1";

// Allow for annotations/options to allow for HTTP transcoding of gRPC requests.
// You can spec & generate these protobufs just fine without, but to actually
// use them, you'll need to use the grpc-gateway plugin to take advantage of
// these annotations
import "google/api/annotations.proto";

service EmployeesService {
  rpc GetEmployee(GetEmployeeRequest) returns (GetEmployeeResponse) {
    // This is where you use those imported annotations above -- this maps GET
    // requests over HTTP to the path specified, for example
    option (google.api.http) = {
      get: "/employees/v1/get_employee/{short_name}"
    };
  }

  rpc ListEmployees(ListEmployeesRequest) returns (ListEmployeesResponse) {
    option (google.api.http) = {
      get: "/employees/v1/list_employees"
    };
  }
}

message Employee {
  int64 id = 1;
  string full_name = 2;
  string birthday = 3;
}

message GetEmployeeRequest {
  string short_name = 1;
}

message GetEmployeeResponse {
  Employee employee = 1;
}

// Even for a Service that you don't have parameters specified for, it's good
// practice to have the parameters be an empty message like this, for
// forwards-compatibility in case you ever do add parameter details
message ListEmployeesRequest {}

message ListEmployeesResponse {
  repeated string short_names = 1;
}
